### PR TITLE
Adds ability to load dynamic configuration objects

### LINKED ISSUE #108: load_configuration does not reset the app viewers
Calling `load_configuration` to dynamically change the viewer config does not reset the existing config.  Rather, it (cumulatively) loads a completely separate configuration, including toolbars and viewers.   

Running
```
app = Application(configuration='cubeviz')
app
app.load_configuration('specviz')
```
produces the following where the new configuration as been loaded to the right of the existing one. 
Screen Shot 2020-05-05 at 12 42 01 PM

### PR DESCRIPTION
**Pull request by havok2063**
This PR enhances the `load_configuration` method to accept a YAML configuration dictionary, in addition to a string name of a preset configuration.  This dictionary is of the same format as the preset configurations.  It includes new methods to `reset_the_configuration`, which fixes #108, and a new method, `get_configuration` which retrieves a preset configuration, e.g. "cubeviz", or "specviz".  This method also allows to extract specific sections of a given configuration.  These methods allow better flexibility and integration of jdaviz into the MAST front-end where MAST needs to dynamically reconfigure, and customize, the application from within a single notebook.  This PR starts to address needs for MAST as indicated in https://jira.stsci.edu/browse/JDAT-460. 

MAST cannot have its configuration in pre-defined YAML files because we need to change the viewers loaded in the application dynamically based on the filename passed to the notebook.  We would need to create multiple YAML configurations files for loading 1d, 2d, 3d files.  Additionally, there may be cases where we need to turn on/off certain plugins for different configurations under certain, more dynamic, circumstances.  This would also involve creating different YAML files for every different use case.    

This PR also includes enhancements to the `app.add_data_to_viewer` and `app.remove_data_from_viewer` methods which enable the user to specify a filename plus an extension to construct a glue data-label, e.g. `app.add_data_to_viewer('flux-viewer', datafile, ext='SCI')`.  This address comments in #112.    

This PR also includes a bare-bones specviz configuration needed to further development for MAST, with the expectation it be superseded by #149.

### CODE REVIEW

**Comment by havok2063** [comment]
I'd be happy to add some documentation on this for the developer section and/or user custom config.

**Comment by javerbukh** [comment]
I tested this and it seems good, I would like to get some other eyes on it before merging. @dtn5ah and @eteq if you guys can review this that would be great! I think the plan will be to wait until after the v1.0 release to merge this as well as a few other PRs. We will try to get these PRs in for v1.1, which will hopefully be released within a sprint or two.

Also, can you rebase and see if that fixes the CI failures?

**Comment by havok2063** [comment]
Thanks for taking a look @javerbukh.  I rebased but it doesn't seem to have helped the tests.

File: jdaviz/core/config.py
Comment (nmearl) [inline jdaviz/core/config.py:11]:
```suggestion
```

File: jdaviz/core/config.py
Comment (nmearl) [inline jdaviz/core/config.py:17]:
```suggestion
```

File: jdaviz/core/config.py
Comment (nmearl) [inline jdaviz/core/config.py:38]: To be consistent, please raise errors instead of using asserts.

File: jdaviz/core/config.py
Comment (nmearl) [inline jdaviz/core/config.py:84]: See above.

File: jdaviz/core/config.py
Comment (nmearl) [inline jdaviz/core/config.py:95]: Minor suggestion.

```suggestion
    """Get a list of pre-built configurations."""
```

File: jdaviz/core/config.py
Comment (havok2063) [reply jdaviz/core/config.py]: Thanks for the comments.  I forgot about the asserts.  This has been fixed.

File: jdaviz/core/config.py
Comment (rosteen) [inline jdaviz/core/config.py:65]: The description here isn't totally accurate, the current configuration retrieval is only in `app.get_configuration`. Seems like the organization of what goes in which `get_configuration` could be a bit more elegant/logically organized but then again I'm not exactly sure how to switch it up.

File: jdaviz/app.py
Comment (havok2063) [reply jdaviz/app.py:1167]: @rosteen No I suppose not.  I originally did not have the `_reset_state` method called by `load_configuration`.  Once I moved it inside, this method is probably redundant now.

File: jdaviz/core/config.py
Comment (havok2063) [reply jdaviz/core/config.py:65]: @rosteen Yeah the docstring here might not be updated.  This code was originally in `app.get_configuration` before I split it out.  I needed to split it out so I could access configurations outside the app context.  The idea behind the new `app.get_configuration` was by default to return the currently loaded one, otherwise return the config specified in the input.

**Review by rosteen — APPROVED** [review]
Looks good to me now, except for one thing. If you start the application with no config loaded and then load a config, the viewers will have headers on them (which currently say "Test" for everything), which is a relic of the default config. I would have thought this wouldn't be the case since the other configs have "tab_headers: False", which I thought referred to this.

I'm inclined to say that this is work that needs to be done on the default config, not something that should hold up this PR. All the functionality I tested switching between and retrieving configs works. Switching to a new config after data has been loaded drops the data, which seems reasonable.

Screen Shot 2020-10-15 at 1 31 49 PM

File: jdaviz/core/config.py
Comment (nmearl) [reply jdaviz/core/config.py:11]: This hasn't been resolved.

File: jdaviz/core/config.py
Comment (nmearl) [reply jdaviz/core/config.py:17]: Same.

File: jdaviz/core/config.py
Comment (nmearl) [reply jdaviz/core/config.py:98]: Same.

**Review by nmearl — CHANGES_REQUESTED** [review]

**Comment by havok2063** [comment]
@rosteen I'm not sure the visible config options work right in general.  For example, in the custom MAST config, I have the visible options to `vis = {'menu_bar': False, 'toolbar': False, 'tray': False, 'tab_headers': False}` for all configs I load (for exaggerated effect), and I still see a tray, menubar, and toolbar.

**Review by javerbukh — APPROVED** [review]
I agree with all the previous comments but I think they can be addressed in another PR. Nice work!!

**Comment by rosteen** [comment]
@nmearl It looks like your outstanding comments were resolved for real this time. Any final objections before I merge this?

**Comment by nmearl** [comment]
@rosteen this looks fine, but the failing tests should be resolved before merging.

**Comment by rosteen** [comment]
Tests passing after a rebase, merging

**Pull request merged**
